README.md
# EbDeployer [![Build Status](https://snap-ci.com/ThoughtWorksStudios/eb_deployer/branch/master/build_image)](https://snap-ci.com/ThoughtWorksStudios/eb_deployer/branch/master) [![Build Status](https://travis-ci.org/ThoughtWorksStudios/eb_deployer.png?branch=master)](https://travis-ci.org/ThoughtWorksStudios/eb_deployer) [![Gem Version](https://badge.fury.io/rb/eb_deployer.svg)](http://badge.fury.io/rb/eb_deployer) [![Dependency Status](https://www.versioneye.com/ruby/eb_deployer/0.6.1/badge.svg)](https://www.versioneye.com/ruby/eb_deployer/0.6.1)

[Built with :yellow_heart: and :coffee: in San Francisco](http://www.thoughtworks.com/mingle/team/)

Low friction deployments should be a breeze. Elastic Beanstalk provides a great foundation for performing Blue-Green deployments, and EbDeployer add a missing link to automate the whole flow out of box.

[Deploy Service by EbDeployer](http://www.thoughtworks.com/mingle/infrastructure/2015/06/01/deploy-service-by-ebdeployer.html)

## Installation

    $ gem install eb_deployer

## Usage

### Step One: AWS Account Setup

Create an AWS IAM user for deploy and give it privilege to operate Elastic Beanstalk. Download the access keys for executing the deployment tasks later. Ensure your command line is primed with appropriate access_keys using one of techniques mentioned on [aws blog](http://ruby.awsblog.com/blog/tag/config). For example using environment variable:

    $ export AWS_ACCESS_KEY_ID=xxx
    $ export AWS_SECRET_ACCESS_KEY=xxx


### Step Two: Packaging

You need to package your application for Elastic Beanstalk stack first. For Java app a warball is appropriate. For Ruby on Rails app a tar.gz file is good. You can also package a Rails/Sinatra app as a war ball using warbler and deploy to Java stack. (Please remember to run rake assets:precompile first for a rails app.)

If you were deploying to Elastic Beanstalk Ruby stack, simply zip your codebase is good enough, for example:

		$ git ls-files | zip your-app-name.zip -@

### Step Three: Generate configuration and Configure deployment process

    $ eb_deploy

This will generate a default configuration at location 'config/eb_deployer.yml'. It is almost empty but working one. And it will generate settings for two environments 'development' and 'production'. Some options can be tweaked. The yml files includes documentation on how you can best suit it to your purpose.


### Step Four: Fasten your seat belt
run deploy

    $ eb_deploy -p <package built> -e <environment>

Then open aws console for Elastic Beanstalk to see the result of this deployment.


### Configure Smoke Testing your stack

EB_Deployer allows you to automate your deployment and then some. You can also add smoke tests to your deployment - thus ensuring that the app you deployed is also working correctly.
Adding a smoke test suite is also simple. Check "smoke_test" section in your eb_deployer.yml. We show a simple curl based smoke test that helps you test if your app is up and responding to http.

    smoke_test: |
      curl_http_code = "curl -s -o /dev/null -w \"%{http_code}\" http://#{host_name}"
      Timeout.timeout(600) do
        until ['200', '302'].include?(`#{curl_http_code}`.strip)
          sleep 5
        end
      end


Any rakeable test suite can be run as part of the smoke test(selenium, cucumber, capybara, and so on.)
You can add more smoke tests by calling arbitrary rake tasks (Please make sure check return status):

    smoke_test: |
      `rake test:smoke HOST_NAME=#{host_name}`
      raise("Smoke failed!") unless $?.success?

Smoke testing gets you one step closer to continuous delivery.

### Blue-Green deployment
Since every deployment now runs smoke test, you now have a better safety net around your deployments. This allows us to trigger automatic blue-green deployments.

To do this you need not do anything special. So far we have deployed the application only once. Let's call this the 'green' stack. Any subsequent calls to deployment will deployment a copy of this application to a new stack - the 'blue' stack. Smoke tests will be run on it and once everything passes the 'blue'(new) stack will be switched to the 'green' stack. Thus your new code will now be on the active stack and the user will experience no downtime.

Once this new stack is stable or has run for a while you can choose to delete the old stack. Or if you are doing continuous delivery you may be ready to another 'blue' deployment. You could just trigger another deployment and repeat this every hour/day/week... you get the idea.



### Destroying a stack
So you are done with this environment, you can destroy it easily as well.

    # destroy one environment
    $ eb_deployer -d -e <environment>


and you are done!

Later tutorials coming soon will cover
* how to setup RDS or other AWS resource and share them between blue green environments

Take a look at code if you can not wait for the documentation.

### eb_deploy.yml configuration file

Like Rails database.yml file, EbDeployer will read eb_deploy.yml file as ERB template file. So you can trade it as standard ERB template file, and substitute Ruby script. For example, the following eb_deployer.yml file partial configures DBPassword parameter for your CloudFormation template from an environment variable:

    resources:
      template: config/my_rds.json
      inputs:
        DBPassword: <%= ENV['MYDBPASSWORD'] %>


EbDeployer also provides the following help methods for convenience:

1. random_hash: it basically is `SecureRandom.hex[0..9]`
2. package_digest: it is your eb package file digest
3. environment: environment name you passed in when executing eb_deploy script using -e option.

## More

[Automated zero downtime deployment to AWS Elastic Beanstalk for Rails with EbDeployer](http://helabs.com/blog/2015/05/19/automated-zero-downtime-deployment-to-aws-elastic-beanstalk-for-rails-with-eb-deployer/?utm_content=buffer12098&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer)

Co-presented session with Amazon ElasticBeanstalk team at AWS re:Invent 2013: [slideshare](http://www.slideshare.net/AmazonWebServices/aws-elastic-beanstalk-under-the-hood-dmg301-aws-reinvent-2013-28428616) and [video (starting at 27:46)](http://www.youtube.com/watch?v=U06-QLd4FL4)

[Deploy Service by EbDeployer](http://www.thoughtworks.com/mingle/infrastructure/2015/06/01/deploy-service-by-ebdeployer.html)

[Rails 3/4 Support](https://github.com/ThoughtWorksStudios/eb_deployer/wiki/Rails-3-Support)

[EbDeployer environment](https://github.com/ThoughtWorksStudios/eb_deployer/wiki/EbDeployer-environment)

[Introduction to EbDeployer](https://www.thoughtworks.com/mingle/scaling/2014/06/13/introduction-to-eb-deployer.html)

[Papertrail Solution for Rails Applications on ElasticBeanstalk
](http://www.thoughtworks.com/mingle/infrastructure/2015/06/10/simple-solution-for-papertrail-on-elasticbeanstalk.html)

[Elastic Beanstalk Tips and Tricks](https://github.com/ThoughtWorksStudios/eb_deployer/wiki/Elastic-Beanstalk-Tips-and-Tricks) (* recommended to read for new Elastic Beanstalk users )

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/ThoughtWorksStudios/eb_deployer/trend.png)](https://bitdeli.com/free "Bitdeli Badge")


CHANGELOG.md
0.7.0
=====

* Add service specific AWS sdk gems
* [#79](https://github.com/ThoughtWorksStudios/eb_deployer/pull/79): Environment updates can complete with errors
* [#82](https://github.com/ThoughtWorksStudios/eb_deployer/pull/82): Added a line to force STDOUT output every line
* [#85](https://github.com/ThoughtWorksStudios/eb_deployer/pull/85): Improve package not found error message
* [#86](https://github.com/ThoughtWorksStudios/eb_deployer/pull/86): Fix to for inactive stack updates when instances are 0
 
0.6.6
=====

* Add support for specifying (and overriding) a [stack policy](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/protect-stack-resources.html) for the CloudFormation resource stack. (contributed by @jlabrecque)
* Fix issue where deployment hangs if resource stack update fails and stack is rolled back, deployment will now fail when the resource stack update fails. (contributed by @jlabrecque)
* Add support for creating environment base on a elasticbeanstalk template. (contributed by @djpate)

0.6.5
=====

* #66: Making it possible to specify other accepted health states when deploying (@dziemid)

0.6.4
=====

* fixes #65 - aws driver should be able to detect CNAME prefixes from both regionalized and legacy EB domains

0.6.3
=====

* Wait until inactive environment status is ready before applying settings.

0.6.0
=====

* Use aws-sdk 2.0
* Ruby 2.2 support
* Robust error handling on CloudFormation resource stack provisioning

0.5.1
=====

* Fix issue: worker tier is picking up old version of aws-sqsd

0.5.0
=====

* Worker tier support

0.4.12
======

* Update default solution stacks.
* Rails generator generates SECRET_KEY_BASE for rails 4 app.

0.4.11
======

* Make gem depend on aws-sdk-v1 until we put in support for both v1 and v2

0.4.10
======

* Fix #33

0.4.9
====

* Ignore AWS::ElasticBeanstalk::Errors::OperationInProgressFailure error when delete application version

0.4.8
====
* Raise an error if the environment launched with problems (contributed by kmanning)
* Add --stack-name option that let's use choose the name of the cloud
  formation stack to operate on (contributed by NET-A-PORTER)
* Document typo/grammar fix (contributed by stig)
* Retry on AWS API throttling error when operating on versions

0.4.7
====
* Added blue-only deployment strategy (a variant of blue-green) that
  skips the cname swap so that the newly deployed code remains on the
  inactive "blue" instance. (contributed by jlabrecque)

0.4.6
====
* Make elasticbeanstalk event polling robust against clock shifting problem.

0.4.5
====
* Rails 3+ support: Rails Generator to install default configurations with
	Postgres RDS resource and everything need for blue-green
	deployment
* Add ability to tag beanstalk environment (from @pmcfadden)

0.4.4
=====
* Fix S3 upload on Windows.
* Experimental support for "inactve_settings" options, which can be used to automatically scale down the inactive environment.
* Update default solution stack name to latest supported JAVA stack.
* Validate solution_stack option before creating environment and list available solution_stacks if not valid.

0.4.3
=====
* Backoff and retry when AWS::ElasticBeanstalk::Errors::Throttling happens during polling events

0.4.2
=====
* Allow provding different deploy strategy for different components.

0.4.1
=====
* Remove options for delete all environments ("-d --all"), because it is too destructive and not recoverable.
* Experimental support for multiple components deployment.

0.4.0
====
* revert back all changes from 0.3.8 to 0.3.9. Elasticbeanstalk haven't relex the real unique constain. The actually contrain is you can not have environment name cross different application

0.3.9
====
* Fix hang problem introduced in 0.3.8 when migrating old ElasticBeanstalk environment.
* Fix issue #13 (Deployment via S3 object broken since 0.3.7)

0.3.8
=====
* Change ElasticBeanstalk environment name pattern. Stop using sufix hash to make eb environment global unique. (Because ElasticBeanstalk does not require environment has globally uniqe name any more.)
* Add migration logic so that if a ElasticBeanstalk environment with legcy name exists, eb_deployer will automaticly terminate it and replace it with ElasticBeanstalk environment has new name pattern.


lib/eb_deployer.rb
require 'digest'
require 'set'
require 'time'
require 'json'
require 'timeout'
require 'aws-sdk-s3'
require 'aws-sdk-elasticbeanstalk'
require 'aws-sdk-cloudformation'
require 'optparse'
require 'erb'
require 'fileutils'

require 'eb_deployer/version'
require 'eb_deployer/utils'
require 'eb_deployer/aws_driver'
require 'eb_deployer/deployment_strategy'
require 'eb_deployer/cloud_formation_provisioner'
require 'eb_deployer/application'
require 'eb_deployer/resource_stacks'
require 'eb_deployer/throttling_handling'
require 'eb_deployer/eb_environment'
require 'eb_deployer/environment'
require 'eb_deployer/default_component'
require 'eb_deployer/component'
require 'eb_deployer/eb_event_source'
require 'eb_deployer/cf_event_source'
require 'eb_deployer/event_poller'
require 'eb_deployer/package'
require 'eb_deployer/config_loader'
require 'eb_deployer/default_config'
require 'eb_deployer/smoke_test'
require 'eb_deployer/version_cleaner'

module EbDeployer

  #
  # Query ouput value of the cloud formation stack
  #
  # @param [String] key CloudFormation output key
  # @param [Hash] opts
  # @option opts [Symbol] :application application name
  # @option opts [Symbol] :environment environment name (e.g. staging, production)
  # @option opts [Symbol] :region AWS Region (e.g. "us-west-2", "us-east-1")
  #
  def self.query_resource_output(key, opts)
    if region = opts[:region]
      Aws.config.update({
        region: region
      })
    end
    app = opts[:application]
    env_name = opts[:environment]
    cf = opts[:cf_driver] || AWSDriver::CloudFormationDriver.new
    stack_name = opts[:stack_name] || "#{app}-#{env_name}"
    provisioner = CloudFormationProvisioner.new(stack_name, cf)
    provisioner.output(key)
  end


  #
  # Deploy a package to specified environments on elastic beanstalk
  #
  # @param [Hash] opts
  #
  # @option opts [Symbol] :application *required* Application name, this
  #   used for isolate packages and contribute to your elastic beanstalk cname
  #   for environments
  #
  # @option opts [Symbol] :environment *required* Environment for same
  #   application, e.g. testing, staging, production. This will map to 2 elastic
  #   beanstalk environments (env-a-xxx, env-b-xxx) if blue-green deployment
  #   strategy specified
  #
  # @option opts [Symbol] :package *required* package for the application
  #   which should be suitable for elastic beanstalk deploying. For example, a
  #   war file should be provided for java solution stacks and a ZIP file
  #   should be provided for Rails or Sinatra stack.
  #
  # @option opts [Symbol] :option_settings  (optional) Elastic Beanstalk
  #   settings that will apply to the environments you deploying. Value should be
  #   array of hash with format such as:
  #
  #     [{
  #        :namespace => 'aws:autoscaling:launchconfiguration',
  #        :option_name => 'InstanceType',
  #        :value => 'm1.small' }]
  #
  #   When there are many, using an external yaml file to hold those
  #   configuration is recommended. Such as:
  #
  #     YAML.load(File.read("my_settings_file.yml"))
  #
  #   For all available options take a look at
  #   http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/command-options.html
  #
  # @option opts [Symbol] :accepted_healthy_states (['Green']) If :accepted_healthy_states
  #   is specified, EBDeployer will accept provided values when checking
  #   health of an environment instead of default value 'Green'. You can use it
  #   to specify additional healthy states, for example: ['Green', "Yellow"]
  #
  # @option opts [Symbol] :phoenix_mode (false) If phoenix mode is turn on, it
  #   will terminate the old elastic beanstalk environment and recreate on
  #   deploy. For blue-green deployment it terminate the inactive environment
  #   first then recreate it. This is useful to avoiding configuration drift and
  #   accumulating state on the EC2 instances. Also it has the benefit of keeping
  #   your EC2 instance system package upto date, because every time EC2 instance
  #   boot up from AMI it does a system update.
  #
  #
  # @option opts [Symbol] :region set the region for application deployment
  #   (e.g. "us-west-2", "us-east-1"). See available zones at
  #   http://aws.amazon.com/elasticbeanstalk/faqs/#regions
  #
  # @option opts [Symbol] :resources If :resources specified, EBDeployer will
  #   use the CloudFormation template you provide to create a default
  #   CloudFormation stack with name <application_name>-<env-name> for the
  #   environment current deploying.  Value of resources need to be hash with
  #   following keys:
  #
  #     :template => CloudFormation template file with JSON format
  #     :policy => CloudFormation policy file with JSON format
  #     :override_policy => (false) If override_policy is true and a policy file is provided then the
  #   policy will temporarily override any existing policy on the resource stack during this update,
  #   otherwise the provided policy will replace any existing policy on the resource stack
  #     :parameters (or :inputs) => A Hash, input values for the CloudFormation template
  #     :transforms => A Hash with key map to your CloudFormation
  #   template outputs and value as lambda that return a single or array of
  #   elastic beanstalk settings.
  #
  #     :capabilities => An array. You need set it to ['CAPABILITY_IAM']
  #
  #   if you want to provision IAM Instance Profile.
  #
  # @option opts [Symbol] :settings See `option_settings`
  #
  # @option opts [Symbol] :package_bucket Name of s3 bucket where uploaded application
  # packages will be stored. Note that the string ".packages" will be added as
  # a suffix to your bucket. So, if "thoughtworks.simple" is passed as the bucket name,
  # the actual s3 bucket name will be thoughtworks.simple.packages.
  #
  # @option opts [Symbol] :smoke_test Value should be a proc or a lambda which
  #   accept single argument that will passed in as environment DNS name. Smoke
  #   test proc or lambda will be called at the end of the deployment for
  #   inplace-update deployment strategy. For blue-green deployment it will run
  #   after inactive environment update finish and before switching.  Defining a
  #   smoke test is high recommended for serious usage. The simplest one could
  #   just be checking the server is up using curl, e.g.
  #
  #     :smoke_test => lambda { |host|
  #       curl_http_code = "curl -k -s -o /dev/null -w \"%{http_code}\" https://#{host}"
  #       Timeout.timeout(600) do
  #         while `#{curl_http_code}`.strip != '200'
  #           sleep 5
  #         end
  #       end
  #     }
  #
  # @option opts [Symbol] :strategy (:blue-green) There are two options:
  #   blue-green or inplace-update. Blue green keep two elastic beanstalk
  #   environments and always deploy to inactive one, to achieve zero downtime.
  #   inplace-update strategy will only keep one environment, and update the
  #   version inplace on deploy. this will save resources but will have downtime.
  #
  # @option opts [Symbol] :solution_stack_name ("64bit Amazon Linux 2013.09 running Tomcat 7 Java 7")
  #   The elastic beanstalk solution stack you want to deploy on top of.
  #
  # @option opts [Symbol] :tier ("WebServer")
  #   The environment tier. Either "WebServer" or "Worker"
  #
  # @option opts [Symbol] :version_label *required*. Version label give the
  #   package uploaded a unique identifier.  Should use something related to
  #   pipeline counter if you have build pipeline setup to build the installer.
  #   For the convenience of dev we recommend use md5 digest of the installer so
  #   that every time you upload new installer it forms a new version. e.g.
  #
  #      :version_label => ENV['MY_PIPELINE_COUNTER']
  #                       || "dev-" + Digest::MD5.file(my_package).hexdigest
  #
  # @options opts [Symbol] :version_prefix. Specifies a prefix to prepend to the
  #   version label. This can be useful if you want to use different binaries for
  #   different environments.
  #
  # @option opts [Symbol] :keep_latest. Specifies the maximum number of versions to
  #   keep.  Older versions are removed and deleted from the S3 source bucket as well.
  #   If specified as zero or not specified, all versions will be kept.  If a
  #   version_prefix is given, only removes version starting with the prefix.
  #
  # @option opts [Symbol] :template_name. Specifies the environement template you wish
  #   to use to build your environment.
  def self.deploy(opts)
    if region = opts[:region]
      Aws.config.update(:region => region)
    end

    bs = opts[:bs_driver] || AWSDriver::Beanstalk.new
    bs = ThrottlingHandling.new(bs, Aws::ElasticBeanstalk::Errors::Throttling)
    s3 = opts[:s3_driver] || AWSDriver::S3Driver.new
    cf = opts[:cf_driver] || AWSDriver::CloudFormationDriver.new

    app_name = opts[:application]
    env_name = opts[:environment]
    version_prefix = opts[:version_prefix].to_s.strip
    version_label = "#{version_prefix}#{opts[:version_label].to_s.strip}"

    application = Application.new(app_name, bs, s3, opts[:package_bucket])
    resource_stacks = ResourceStacks.new(opts[:resources],
                                         cf,
                                         !!opts[:skip_resource_stack_update],
                                         opts[:tags])

    stack_name = opts[:stack_name] || "#{app_name}-#{env_name}"

    environment = Environment.new(application, env_name, stack_name, bs) do |env|
      env.resource_stacks = resource_stacks
      env.settings = opts[:option_settings] || opts[:settings] || []
      env.inactive_settings = opts[:inactive_settings] || []
      env.creation_opts = {
        :template_name => opts[:template_name],
        :solution_stack => opts[:solution_stack_name],
        :cname_prefix =>  opts[:cname_prefix],
        :smoke_test => opts[:smoke_test],
        :phoenix_mode => opts[:phoenix_mode],
        :accepted_healthy_states => opts[:accepted_healthy_states],
        :blue_green_terminate_inactive => opts[:blue_green_terminate_inactive] || false,
        :blue_green_terminate_inactive_wait => opts[:blue_green_terminate_inactive_wait] || 600,
        :blue_green_terminate_inactive_sleep => opts[:blue_green_terminate_inactive_sleep] || 15,
        :tags => opts[:tags],
        :tier => opts[:tier]
      }
      env.strategy_name = opts[:strategy] || :blue_green
      env.components = opts[:components]
      env.component_under_deploy = opts[:component]
    end

    application.create_version(version_label, opts[:package])
    environment.deploy(version_label)
    application.clean_versions(version_prefix, opts[:keep_latest].to_i || 0)
  end

  def self.destroy(opts)
    if region = opts[:region]
      Aws.config.update(:region => region)
    end

    app = opts[:application]
    bs = opts[:bs_driver] || AWSDriver::Beanstalk.new
    s3 = opts[:s3_driver] || AWSDriver::S3Driver.new

    Application.new(app, bs, s3).delete(opts[:environment])
  end

  def self.cli
    options = {
      :action => :deploy,
      :environment => 'dev',
      :config_file => 'config/eb_deployer.yml'
    }

    parser = cli_parser(options)
    parser.parse!
    action = options.delete(:action)

    if File.exist?(options[:config_file])
      puts "Found configuration at #{options[:config_file]}."
    else
      puts "Generated default configuration at #{options[:config_file]}."
      DefaultConfig.new(File.basename(Dir.pwd)).write_to(options[:config_file])
      exit(2)
    end

    if !options[:package] && action == :deploy
      puts "Missing options: -p (--package)"
      puts "'eb_deploy --help' for details"
      puts parser
      exit(-1)
    end

    self.send(action, ConfigLoader.new.load(options))
  end

  private

  def self.cli_parser(options)
    OptionParser.new do |opts|
      opts.banner = "Usage: eb_deployer [options]"
      opts.on("-p", "--package [FILE/S3_OBJECT]", "Package to deploy, can be a war file for java application, or yaml specification for package location on S3, or a S3 object & bucket name separated by colon, e.g. bucket_name:key_name") do |v|
        options[:package] = v
      end

      opts.on("-e", "--environment [ENV_NAME]", "(Defaults to 'dev') Environment on which to operate (e.g. dev, staging, production). This must be defined in 'environments' section of the config file") do |v|
        options[:environment] = v
      end

      opts.on("-c", "--config-file [FILE]", "eb_deployer config file. Default location is config/eb_deployer.yml") do |v|
        options[:config_file] = v
      end

      opts.on("-d", "--destroy", "Destroy all Elasticbeanstalk environments under the application which have specified environment as name prefix") do |v|
        options[:action] = :destroy
      end

      opts.on("-s", "--stack-name [STACK_NAME]", "CloudFormation stack name to use. If not specified, defaults to {app}-{env_name}") do |v|
        options[:stack_name] = v
      end

      opts.on("--skip-resource-stack-update", "Skip cloud-formation stack update. (only for extreme situation like hitting a cloudformation bug)") do |v|
        options[:skip_resource_stack_update] = true
      end

      opts.on("--component [COMPONENT]", "Specify which component to deploy") do |v|
        options[:component] = v
      end

      opts.on("-v", "--version", "Print current version") do |v|
        puts "eb_deployer v#{VERSION}"
        exit(0)
      end

      opts.on("--debug", "Output AWS debug log") do |d|
        require 'logger'
        logger = Logger.new($stdout)
        logger.level = Logger::DEBUG
        Aws.config.update(:logger => logger)
      end

      opts.on("-h", "--help", "help")  do
        puts opts
        puts ""
        puts "S3 object package format: s3_bucket_name:s3_object_key"
        puts "YAML package file format:"
        puts "s3_bucket: <bucket_name>"
        puts "s3_key: <object_path>"
        exit(0)
      end
    end
  end


end
